crates/red_knot_python_semantic/src/types.rs

```diff
@@ -981,7 +987,10 @@ impl<'db> Type<'db> {
             }
             (Type::SubclassOf(subclass_of_ty), Type::Instance(_))
                 if subclass_of_ty.is_dynamic()
-                    && target.is_assignable_to(db, KnownClass::Type.to_instance(db)) =>
+                    && (target.is_assignable_to(db, KnownClass::Type.to_instance(db))
+                        || KnownClass::Type
+                            .to_instance(db)
+                            .is_assignable_to(db, target)) =>
             {
                 true
             }
```

Comment:
I've spent a while staring at this branch now 😆 I _think_ it's correct, it just took me a little bit of time to wrap my head around it -- some comments would be good. It's saying that `type[Any]` is a consistent subtype of a type `T` if `T` is either a consistent subtype or a consistent supertype of `type`. Right?

Reply:
> I've spent a while staring at this branch now

So glad I'm not the only one. I attributed it to tiredness :smile:

Reply:
I also found it confusing, both in its form prior to this change, and still after this change :) I think it's correct, though. I'll try to formulate a clear argument for why it is correct, and put that in a comment.

Reply:
I pushed comments for every case in `is_assignable_to`, including this one.

I found it pretty easy to justify the new clause I added here in this PR; `type[Any]` is assignable to anything `type[object]` is assignable to, because `type[Any]` can materialize to `type[object]`.

What I found harder to formulate a justification for was the previous clause that was already here, which said that `type[Any]` is assignable to anything which is assignable to `type[object]`. I found it clearer if I tightened that to "`type[Any]` is assignable to anything which is a subtype of `type[object]`", and that form still passes all tests, including property tests.